examples/logisticregression/main.go

```diff
@@ -1,185 +1,288 @@
+	// create a new graph and add x, y, w, b, and one as Nodes
+	g := G.NewGraph()
+	x := G.NewMatrix(g, Float, G.WithName("x"), G.WithShape(N, feats))
+	y := G.NewVector(g, Float, G.WithName("y"), G.WithShape(N))
+
+	w := G.NewVector(g, Float, G.WithName("w"), G.WithShape(feats))
+	b := G.NewScalar(g, Float, G.WithName("bias"))
+	// Add a constant node 1 to be used later in the loss function
+	one := G.NewConstant(1.0)
+
+	// Here we create the nodes that will do the prediction operations
+	// create a node that has the operation: (x*w + b)
+	xwmb := G.Must(G.Add(G.Must(G.Mul(x, w)), b))
+	// create a node that has the operation: sigmoid(xwmb)
+	prob := G.Must(G.Sigmoid(xwmb))
+	G.WithName("prob")(prob)
+	// create a "pred" node that has the operation (prod > 0.5)
+	// this ensures that our prediction output is {0,1}
+	pred := G.Must(G.Gt(prob, G.NewConstant(0.5), false))
+	G.WithName("pred")(pred)
+
+	// Here we create the nodes that contain the operations that
+	// will calculate the cost function.
+	// binary cross entropy: -y * log(prob) - (1-y)*log(1-prob)
+	logProb := G.Must(G.Log(prob))
+	fstTerm := G.Must(G.HadamardProd(G.Must(G.Neg(y)), logProb))
+	oneMinusY := G.Must(G.Sub(one, y))
+	logOneMinusProb := G.Must(G.Log(G.Must(G.Sub(one, prob))))
+	sndTerm := G.Must(G.HadamardProd(oneMinusY, logOneMinusProb))
+
+	crossEntropy := G.Must(G.Sub(fstTerm, sndTerm))
+	G.WithName("crossEntropy")(crossEntropy)
+	loss := G.Must(G.Mean(crossEntropy))
+
+	// regularization term
+	weightSq := G.Must(G.Square(w))
+	sumSq := G.Must(G.Sum(weightSq))
+	l1reg := G.NewConstant(0.01, G.WithName("l1reg"))
+	regTerm := G.Must(G.Mul(l1reg, sumSq))
+
+	// cost we want to minimize
+	cost := G.Must(G.Add(loss, regTerm))
+	G.WithName("cost")(cost)
+
+	// calculate gradient by backpropagation https://en.wikipedia.org/wiki/Backpropagation
+	grads, err := G.Grad(cost, w, b)
+	handleError(err)
+	// "dcost/dw" == derivative of cost with respect to w
+	G.WithName("dcost/dw")(grads[0])
+	// "dcost/db" == derivative of cost with respect to b
+	G.WithName("dcost/db")(grads[1])
+
+	// create the nodes for calculating the gradient
+	learnRate := G.NewConstant(0.1) // be careful not to set a learnRate too high
+	gwlr := G.Must(G.Mul(learnRate, grads[0]))
+	wUpd := G.Must(G.Sub(w, gwlr))
+	gblr := G.Must(G.Mul(learnRate, grads[1]))
+	bUpd := G.Must(G.Sub(b, gblr))
+
+	// create the nodes that do the final operation of updating w and b
+	G.Set(w, wUpd)
+	G.Set(b, bUpd)
+	// write to the gographviz file for debugging https://github.com/awalterschulze/gographviz
+	ioutil.WriteFile("fullGraph.dot", []byte(g.ToDot()), 0644)
+
+	// Now that we have created all the notes, we can compile a program.
+	// We are essentially creating a list of instructions to get from our inputs {x, y}
+	// to our outputs {wUpd, bUpd}. Note that we only need to create the progam once.
+	prog, locMap, err := G.CompileFunction(g, G.Nodes{x, y}, G.Nodes{wUpd, bUpd})

```

Comment:
This is a hole in my understanding of gorgonia. I thought we had to compile a program before running it. Since that's not the case, how would the tape machine know which node to start from? What would the complete function call look like?

Reply:
> I thought we had to compile a program before running it.

With the `tapeMachine` yes. You have to do that. The standard way is creating a `tapeMachine` by using `machine := NewTapeMachine(g)`. Internally, Gorgonia will figure out that a compilation process needs to happen, so it will call `Compile(g)`, which will output the `prog` and `locMap` internally.

How it knows where to start from is simple: when nodes are added to the graph to Gorgonia, nodes that have no precedents and have a nil `Op` are marked as input nodes. 

Optionally, one may precompile the graph first, yielding a pattern that looks like that:

```
prog, locmap, err := Compile(g)
machine := NewTapeMachine(g, WithPrecompiled(prog, locmap))
```

When would someone need to do that? Well, I personally recommend starting out like that before using the shortcut version above. In fact if you reference older material on Gorgonia, you will note that the shortcut method was once there, removed and then readded. 

Another time one might want to use a `WithPrecompiled` option is when you have a large graph, and just want to compile a subset of it. In the logisticregression example, because we've put TWO things into the same graph (pred and train), you just want to compile only the training portion of it.

To do so, you define what the inputs and outputs are:

```
prog, locmap, err := CompileFunction(g, INPUTNODES, OUTPUTNODES)
```

in my opinion only very optimized programs should do this.